### PR TITLE
[Cpp API Compatibility] Add Tensor::index_reduce compat interface

### PR DESCRIPTION
**Pull request by youge325**
### PR Category
Execute Infrastructure

### PR Types
New features

### Description
Add PyTorch-compatible `Tensor::index_reduce` / `Tensor::index_reduce_` APIs in the Paddle C++ compat layer.

- New compat header `ATen/ops/index_reduce.h`: maps `index_reduce` → `paddle::experimental::put_along_axis`
  - Reduce name mapping: `"prod"` → `"multiply"`
  - Index shape adaptation: PyTorch uses 1D index while Paddle `put_along_axis` requires index matching source dimensions, auto-converted via `reshape` + `expand`
- New `c10/util/string_view.h`: defines `c10::string_view` as alias to `std::string_view`
- Add `Tensor::index_reduce` / `Tensor::index_reduce_` declarations in `ATen/core/TensorBody.h`
- Add include in `ATen/Functions.h`
- Add `test/cpp/compat/ATen_index_reduce_test.cc` with 13 gtest cases covering:
  - prod / mean / amax / amin reduce modes
  - include_self=true/false
  - In-place (`index_reduce_`) and out-of-place variants
  - Non-member function `at::index_reduce`
  - dim=0 and dim=1
  - 1D tensor boundary case
  - Float / Double / Int / Long dtypes
  - Invalid reduce type exception

### 是否引起精度变化
否

### CODE REVIEW

File: paddle/phi/api/include/compat/ATen/ops/index_reduce.h [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+// PyTorch index_reduce uses a 1D index where each entry maps a slice of source
+// to a slice of self along the operation dimension. Paddle put_along_axis
+// requires index and source to have matching shape (or broadcastable).
+// This helper broadcasts the 1D index across all non-operation dimensions.
+inline paddle::Tensor expand_index_for_put_along_axis(
+    const paddle::Tensor& index, const paddle::Tensor& source, int dim) {
+  if (index.dims().size() == 1 && source.dims().size() > 1) {
+    // Reshape index: [N] -> [1, ..., N, ..., 1] where N is at dim position
+    std::vector<int64_t> reshape_shape(source.dims().size(), 1);
+    reshape_shape[dim] = index.dims()[0];
```
Comment (risemeup1111) [inline paddle/phi/api/include/compat/ATen/ops/index_reduce.h:119]: P1 **优先级：P1**

这里直接用 `dim` 作为 `reshape_shape` 下标会让合法的 PyTorch 调用 `index_reduce(-1, ...)` 写到 `reshape_shape[-1]`；`dim >= rank` 时也会先发生越界访问，而不是抛出可控错误。这个 API 是做 PyTorch 兼容的，负维度是常见用法，请在进入 reshape/put_along_axis 前统一把 `dim` 规范化并做范围、rank 和 index 向量校验，然后所有 out/in-place 变体都传规范化后的正维度。

建议形状如下，并补一个 `dim=-1` 的回归测试：

```cpp
inline int64_t normalize_index_reduce_dim(const at::Tensor& self,
                                          const at::Tensor& index,
                                          const at::Tensor& source,
                                          int64_t dim) {
  const int64_t ndim = self.dim();
  if (dim < 0) dim += ndim;
  PD_CHECK(dim >= 0 && dim < ndim, "index_reduce(): dim out of range");
  PD_CHECK(source.dim() == ndim, "index_reduce(): source rank must match self");
  PD_CHECK(index.dim() == 1, "index_reduce(): index must be a 1-D tensor");
  PD_CHECK(index.size(0) == source.size(dim),
           "index_reduce(): index length must match source.size(dim)");
  for (int64_t i = 0; i < ndim; ++i) {
    if (i != dim) {
      PD_CHECK(source.size(i) == self.size(i),
               "index_reduce(): source dimensions other than dim must match self");
    }
  }
  return dim;
}

const auto pd_dim = static_cast<int>(
    normalize_index_reduce_dim(self, index, source, dim));
auto pd_index = expand_index_for_put_along_axis(
    index._PD_GetInner(), source._PD_GetInner(), pd_dim);
// put_along_axis(..., pd_dim, ...)
```

File: paddle/phi/api/include/compat/ATen/ops/index_reduce.h [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+#include "paddle/phi/api/include/api.h"
+
+namespace at {
+
+// Helper: convert PyTorch reduce name to Paddle reduce name
+inline std::string convert_reduce_name(c10::string_view reduce) {
+  if (reduce == "prod") {
+    return "multiply";
+  }
+  return std::string(reduce);
```
Comment (risemeup1111) [inline paddle/phi/api/include/compat/ATen/ops/index_reduce.h:37]: P1 **优先级：P1**

这里把未知 reduce 原样透传给 `put_along_axis`，会把 Paddle 支持但 PyTorch `Tensor::index_reduce` 不支持的值静默接受，例如 `"add"`、`"assign"`、`"mul"`/`"multiply"`。这会让迁移代码在 Paddle compat 层下通过并产生结果，但在 PyTorch 下应当报错，属于兼容层行为偏差。请在 compat API 层显式限制为 PyTorch 的合法集合，并只把 `prod` 映射到 Paddle 的 `multiply`。

```suggestion
  PD_THROW(
      "index_reduce(): reduce must be one of 'prod', 'mean', 'amax' or 'amin', but got '",
      reduce,
      "'.");
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
发现需要先修正的兼容性问题，详细建议已放在 inline review comments 中。主要风险是 PyTorch 兼容 API 对合法负维度和非法参数的处理与预期不一致，建议补齐相应回归测试后再继续合入。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by youge325** [comment]
@PaddlePaddle-bot @risemeup1111 感谢 review，所有问题已修复：

1. **index_reduce_out 7参数调用** → 改为先调用 put_along_axis 获取结果，再赋值给 out
2. **convert_reduce_name 缺少 sum 映射** → 已添加 sum → add 映射
3. **负维度处理** → 新增 normalize_index_reduce_dim() 辅助函数，统一规范化 dim（处理负值、范围检查、rank/index/shape 校验），所有 out/in-place 变体均使用规范化后的正维度
4. **未知 reduce 透传** → 在 compat 层显式限制 reduce 为 PyTorch 合法集合 {prod, sum, mean, amax, amin}，非法值立即抛异常
5. **补 dim=-1 回归测试** → 新增 SumBasic2D 和 NegativeDim 测试用例

本地验证：Paddle 编译通过，ctest (ATen|c10|torch) 69/69 通过，PCAT result_cmp MATCH。

File: paddle/phi/api/include/compat/ATen/ops/index_reduce.h [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+           "index_reduce(): dim out of range (expected to be in range of [",
+           -ndim,
+           ", ",
+           ndim - 1,
+           "], but got ",
+           dim - ndim,
+           ")");
+  PD_CHECK(source.dim() == ndim,
+           "index_reduce(): source rank must match self rank");
+  PD_CHECK(index.dim() == 1, "index_reduce(): index must be a 1-D tensor");
```
Comment (risemeup1111) [inline paddle/phi/api/include/compat/ATen/ops/index_reduce.h:63]: P1 **优先级：P1**

这里仍缺少 dtype 级别的兼容校验。PyTorch 的 `index_reduce` 要求 `self` 和 `source` dtype 一致，`index` 只能是 int32/int64；当前代码只校验 rank/shape，float index 会继续传到 Paddle `put_along_axis`，而 CPU/GPU kernel 只处理 INT32/INT64 分支，坏 dtype 可能直接返回未 reduce 的拷贝或在更深处报不清晰的错误。请在 compat 层提前拦截，并补充非法 index/source dtype 的回归测试。

```suggestion
  PD_CHECK(source.dim() == ndim,
           "index_reduce(): source rank must match self rank");
  PD_CHECK(self.scalar_type() == source.scalar_type(),
           "index_reduce(): self and source must have the same scalar type");
  PD_CHECK(index.dim() == 1, "index_reduce(): index must be a 1-D tensor");
  PD_CHECK(index.scalar_type() == at::kLong || index.scalar_type() == at::kInt,
           "index_reduce(): Expected dtype int32/int64 for index but got ",
           index.scalar_type());
```

File: paddle/phi/api/include/compat/ATen/ops/index_reduce.h [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+  auto pd_dim = normalize_index_reduce_dim(self, index, source, dim);
+  auto pd_index = expand_index_for_put_along_axis(
+      index._PD_GetInner(), source._PD_GetInner(), pd_dim);
+  auto result = paddle::experimental::put_along_axis(self._PD_GetInner(),
+                                                     pd_index,
+                                                     source._PD_GetInner(),
+                                                     static_cast<int>(pd_dim),
+                                                     pd_reduce,
+                                                     include_self);
+  out._PD_GetInner() = result;
```
Comment (risemeup1111) [inline paddle/phi/api/include/compat/ATen/ops/index_reduce.h]: P1 **优先级：P1**

`index_reduce_out` 这里把 `out` wrapper 内部的 `paddle::Tensor` 句柄替换成了新结果，而不是写入调用方传入的 pre-allocated out 存储。这样 `auto* p = out.data_ptr<T>(); at::index_reduce_out(out, ...);` 之后，旧的 out 存储不会被更新，out 的别名/共享引用也观察不到结果；同目录现有 `sum_out` 是先算结果再用 `assign_out_` 写回 out。请改成写入 out 持有的目标 tensor，而不是重新绑定 wrapper。

```suggestion
  paddle::experimental::assign_out_(result, out._PD_GetInner());
```

File: paddle/phi/api/include/compat/ATen/ops/index_reduce.h (hunk 2 above)
Comment (risemeup1111) [reply paddle/phi/api/include/compat/ATen/ops/index_reduce.h:37]: 补充核对后，这个线程里的白名单问题还没有完全解决。当前实现新增了 `"sum" -> "add"`，但 PyTorch `Tensor.index_reduce_` 文档和 native meta 检查仍只允许 `"prod"`, `"mean"`, `"amax"`, `"amin"`；`"sum"` 属于 `scatter_reduce` 的 reduce 取值，不是 `index_reduce` 的合法取值。请把 `sum` 映射和对应正向测试移除，改成 `ASSERT_THROW(..., "sum")` 之类的回归测试，避免 compat 层接受 PyTorch 会拒绝的调用。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交：之前的负维度越界问题基本已修复，但 reduce 白名单仍与 PyTorch `index_reduce` 不一致；另外本轮发现 dtype 校验和 out 变体写回语义还有需要修正的问题。具体位置和建议已放在 inline review comments 及原有线程回复中。

当前 CI 仍有部分构建/测试在排队，尚未全部完成。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by youge325** [comment]
@risemeup1111 已按 review 意见完成第二轮修复：

1. **reduce 白名单** — 移除 "sum" 映射，白名单改为 {prod, mean, amax, amin}，与 PyTorch index_reduce 文档一致；新增 SumInvalidThrows 回归测试验证 "sum" 会抛异常。

2. **dtype 校验** — 在 normalize_index_reduce_dim() 中新增：
   - self.scalar_type() == source.scalar_type()
   - index.scalar_type() 为 kLong 或 kInt
   新增 IndexDtypeInvalidThrows 和 SourceDtypeMismatchThrows 回归测试。

3. **index_reduce_out 写回语义** — 改为 paddle::experimental::assign_out_(result, out._PD_GetInner())，写入 out 持有的目标存储而非替换 wrapper 句柄。

本地验证：Paddle 编译通过，ctest 69/69 通过，PCAT result_cmp MATCH。

**Review by risemeup1111 — APPROVED** [review]
已复查最新提交，作者提到的第二轮修复已在当前 diff 中体现：reduce 白名单已去掉 `sum`，dtype 校验已补齐，`index_reduce_out` 也改为写回 `out` 持有的目标存储。前一轮 Nyanpasu 提出的阻塞问题看起来已解决，当前未发现新的需要阻塞合入的问题。

CI 仍有部分任务在进行中，请以最终 CI 结果为准。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddle/phi/api/include/compat/ATen/ops/index_reduce.h [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+           ", ",
+           ndim - 1,
+           "], but got ",
+           original_dim,
+           ")");
+  PD_CHECK(source.dim() == ndim,
+           "index_reduce(): source rank must match self rank");
+  PD_CHECK(self.scalar_type() == source.scalar_type(),
+           "index_reduce(): self and source must have the same scalar type");
+  PD_CHECK(index.dim() == 1, "index_reduce(): index must be a 1-D tensor");
```
Comment (risemeup1111) [inline paddle/phi/api/include/compat/ATen/ops/index_reduce.h:63]: P1 **优先级：P1**

这里把 `index` 限制成严格 1-D，会拒绝 PyTorch 合法的标量 index。PyTorch 的 `index_reduce` meta 检查允许 `index.dim() <= 1`，例如 `torch.ones(3).index_reduce(0, torch.tensor(1), torch.tensor([2.]), "prod")` 会成功并只更新位置 1；当前 compat 层会在这里直接抛错，导致迁移代码里单个 index 的调用无法对齐。

请把 0-D index 按长度 1 处理，并在传给 `put_along_axis` 前 reshape 成 `{1}` 后再按现有逻辑扩展到 `source` 形状。修复形状可以是：

```cpp
PD_CHECK(index.dim() <= 1,
         "index_reduce(): index is supposed to be a vector");
PD_CHECK(index.numel() == source.size(dim),
         "index_reduce(): index length must match source.size(dim)");

// In expand_index_for_put_along_axis:
paddle::Tensor normalized_index = index;
if (normalized_index.dims().size() == 0) {
  normalized_index = paddle::experimental::reshape(
      normalized_index, paddle::experimental::IntArray({1}));
}
// Then use normalized_index in the existing reshape/expand path.
```

同时补一个 scalar index 的回归测试，覆盖 out-of-place 和 in-place 至少一个路径。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查当前提交，发现一个仍需先修正的 PyTorch 兼容行为缺口：`index_reduce` 当前会拒绝合法的标量 `index`。详细原因和建议已放在 inline review comment 中。

当前 CI 仍有部分任务在排队或运行，请以最终 CI 结果为准。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1 — APPROVED** [review]
